### PR TITLE
feat(core): reimplement blocking workflow updates on interim changes

### PR DESCRIPTION
**Pull request by ivov**
Alternative to https://github.com/n8n-io/n8n/pull/4397

### CODE REVIEW

File: packages/cli/src/databases/entities/WorkflowEntity.ts [hunk 1]
```diff
@@ -84,6 +88,30 @@ export class WorkflowEntity extends AbstractEntity implements IWorkflowDb {
...
+
+		const state = JSON.stringify({
+			name,
+			active,
+			nodes,
+			connections,
+			settings,
+			staticData,
+			pinData,
+		});
```
Comment (krynble) [inline packages/cli/src/databases/entities/WorkflowEntity.ts:111]: I think we should simplify this to use only nodes and connections. Maybe pinData.

This reduces the chance of having issues with unnecessary warnings such as name changes or activation status.

File: packages/cli/src/workflows/workflows.controller.ts [hunk 2]
```diff
@@ -355,6 +355,14 @@ workflowsController.patch(
 			);
 		}
 
+		if (incomingHash !== shared.workflow.hash) {
```
Comment (krynble) [inline packages/cli/src/workflows/workflows.controller.ts:358]: Can we make this check only if the payload contains `nodes` and `connections`? Perhaps `pinData` too?

This reduces the chance of us making checks and warnings unnecessarily.

**Review by krynble — DISMISSED** [review]
Looking great, tested with two users, multiple changes, all working fine. My two comments aim to simplify the checks and reduce possibly unnecessary warnings.

File: packages/cli/src/databases/entities/WorkflowEntity.ts (hunk 1 above)
Comment (ivov) [reply packages/cli/src/databases/entities/WorkflowEntity.ts:111]: Thanks! As discussed, any editable parts of the workflow that we exclude from the hash will cause changes to those parts to be overwritable, which we likely do not want, so keeping it strict for now.

File: packages/cli/src/workflows/workflows.controller.ts (hunk 2 above)
Comment (ivov) [reply packages/cli/src/workflows/workflows.controller.ts:358]: https://github.com/n8n-io/n8n/pull/4446#discussion_r1006571003

**Review by krynble — APPROVED** [review]

**Comment by janober** [comment]
Got released with `n8n@0.201.0`

**Pull request merged**